crates/ra_editor/src/col_index.rs

```diff
@@ -0,0 +1,256 @@
+//! A data structure to map utf-8 based columns to utf-16 based ones
+
+use rustc_hash::FxHashMap;
+use crate::{LineCol, TextUnit};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ColIndex {
+    utf16_lines: FxHashMap<u32, Vec<Utf16Char>>,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+struct Utf16Char {
+    start: TextUnit,
+    end: TextUnit,
+}
+
+impl Utf16Char {
+    fn len(&self) -> TextUnit {
+        self.end - self.start
+    }
+}
+
+impl ColIndex {
+    pub fn new(text: &str) -> ColIndex {
+        let mut utf16_lines = FxHashMap::default();
+        let mut utf16_chars = Vec::new();
+        let mut line = 0;
+        let mut curr = 0.into();
+        for c in text.chars() {
+            if c == '\n' {
+                // Save any utf-16 characters seen in the previous line
+                if utf16_chars.len() > 0 {
+                    utf16_lines.insert(line, utf16_chars);
+                    utf16_chars = Vec::new();
+                }
+
+                // Prepare for processing the next line
+                curr = 0.into();
+                line += 1;
+                continue;
+            }
+
+            let char_len = TextUnit::of_char(c);
+            if char_len.to_usize() > 1 {
+                utf16_chars.push(Utf16Char {
+                    start: curr,
+                    end: curr + char_len,
+                });
+            }
+
+            curr += char_len;
+        }
+        ColIndex { utf16_lines }
+    }
+
+    pub fn utf8_to_utf16_col(&self, mut line_col: LineCol) -> LineCol {
```

Comment:
Hm, this `LineCol` -> `LineCol` API seems a bit error prone, because it uses the same type for different units of measure. This can lead to errors: just yesterday my bank showed me the amount of money on my account in rubles, while using euro as a currency sign :D

I think a lower-level API might be safer:

```
pub fn col_as_utf16(&self, line_col: LineCol) -> usize {...}
```

Note that, by definition, `TextUnit` is *always* a `utf_8` length, so using it for utf-16 is not correct.

Reply:
Ok, then I will change the type of `col` in `LineCol`, since it is being used for UTF16 indices